XMonad/Hooks/EwmhDesktops.hs

```diff
@@ -320,10 +320,28 @@ ewmhDesktopsLogHook' EwmhDesktopsConfig{workspaceSort, workspaceRename} = withWi
     let clientList = nub . concatMap (W.integrate' . W.stack) $ ws
     whenChanged (ClientList clientList) $ setClientList clientList
 
-    -- Set stacking client list which should have bottom-to-top
-    -- stacking order, i.e. focused window should be last
-    let clientListStacking = nub . concatMap (maybe [] (\(W.Stack x l r) -> reverse l ++ r ++ [x]) . W.stack) $ ws
-    whenChanged (ClientListStacking clientListStacking) $ setClientListStacking clientListStacking
+    -- @ws@ is sorted in the "workspace order", which, by default, is
+    -- the lexicographical sorting on @WorkspaceId@.
+    -- @_NET_CLIENT_LIST_STACKING@ is expected to be in the "bottom-to-top
+    -- stacking order".  It is unclear what that would mean for windows on
+    -- invisible workspaces, but it seems reasonable to assume that windows on
+    -- the current workspace should be "at the top".  With the focused window to
+    -- be the top most, meaning the last.
+    let clientListStacking =
+          let wsInFocusOrder = W.hidden s
+                               ++ (map W.workspace . W.visible) s
+                               ++ [W.workspace $ W.current s]
+              stackWindows (W.Stack cur up down) = down ++ reverse up ++ [cur]
+              workspaceWindows = maybe [] stackWindows . W.stack
+              -- In case a window is a member of multiple workspaces, we keep
+              -- only the last occurrence in the list.  One that is closer to
+              -- the top in the focus order.
+              uniqueKeepLast ls = [ e | (i, e) <- zip [0..] ls
+                                      , notElem e $ take i ls
+                                  ]
```

Comment:
This is not correctly implemented. 
```haskell
ghci> uniqueKeepLast [1,3,1]
[1,3]
```


```suggestion
              uniqueKeepLast = reverse . nub . reverse
```

Reply:
You are right.  I should have checked it.
Changed to

```haskell
              uniqueKeepLast ls = [ e | (i, e) <- zip [0..] ls
                                      , notElem e $ drop (i + 1) ls
                                  ]
```

But if you think that `reverse . nub . reverse` is a better choice, I'll change it.

Reply:
`reverse . nub . reverse` is more readable for most Haskellers, and not any slower, so I'd say it's preferable.

Reply:
Got it.  Changed as you suggested.

Reply:
Not to bikeshed this too much, but since we are talking about _all_ windows here, it may be advisable to use [nubOrd](https://hackage.haskell.org/package/containers-0.6.6/docs/Data-Containers-ListUtils.html#v:nubOrd)

Reply:
Oh, didn't even realize that one is in the standard library! (But it probably won't be for the entire range of stackage versions we support...)

Reply:
Ah, indeed, GHC 8.4.4 still ships with `containers` version 0.5.11.0, which is the last version that does not have this...

Reply:
We could possibly stop supporting 8.4, as Debian stable has 8.8 and Ubuntu 20.04 has 8.6, and those were the benchmark I recall.

Reply:
Or just define it ourselves, since it's dead simple (`Set.toList . Set.fromList`).

Reply:
That would destroy the order of things, right?  `nubOrd` in containers is defined in a way so that doesn't happen; something like

``` haskell
nubOrd :: Ord b => [a] -> [a]
nubOrd = go Set.empty
 where
  go _ [] = []
  go s !(!x : xs)
    | x `Set.member` s =     go s                 xs
    | otherwise        = x : go (Set.insert fx s) xs
```

Maybe `nub` is fine here and this really is bikeshedding, I don't know (I remember seeing improvements starting only at about 20 or 30 elements)

Reply:
Pretty much, yes. I know I've told people that our main use of `nub` is fine because if you have enough screens for `nubOrd` to show an improvement then you have bigger problems than `nub`…

Reply:
I do not have an opinion here, but I can make the change, if `nubOrd` is desired.
I could either remove `8.4.4` from the tested list, and bump both `base` and `containers` minimal version.
Or I can put `nubOrd` implementation somewhere.
It seems that the library implementation is a bit more complex, as it supports fusion.
But I can also implement it just for this case at the use site, calling it `uniqueKeepLast`.